src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/CollectTraceTests.cs

```diff
@@ -0,0 +1,155 @@
+            {
+                { "message", TestAppScenarios.TraceEvents.UniqueEventMessage },
+                { "foobar", "baz" }
+            }));
+        }
+
+        [Fact]
+        public Task StopOnEvent_DoesNotStop_WhenPayloadFieldValueMismatch()
+        {
+            return Assert.ThrowsAsync<TaskCanceledException>(() => StopOnEventTestCore(payloadFilter: new Dictionary<string, string>()
+            {
+                { "message", TestAppScenarios.TraceEvents.UniqueEventMessage.ToUpperInvariant() }
+            }));
+        }
+
+        [Fact]
+        public Task StopOnEvent_UsesDuration_WhenNoEventMatchesInTime()
+        {
+            return StopOnEventTestCore(opcode: TraceEventOpcode.Resume, duration: TimeSpan.FromSeconds(10));
+        }
+
+        private async Task StopOnEventTestCore(TraceEventOpcode opcode = ExpectedEventOpcode, IDictionary<string, string> payloadFilter = null, TimeSpan? duration = null)
+        {
+            const string DefaultRuleName = "FunctionalTestRule";
+            const string EgressProvider = "TmpEgressProvider";
+            const string EventProviderName = "TestScenario";
+            const string StoppingEventName = "UniqueEvent";
+
+            string qualifiedEventName = (opcode == TraceEventOpcode.Info) ? StoppingEventName : FormattableString.Invariant($"{StoppingEventName}/{opcode}");
+
+            using TemporaryDirectory tempDirectory = new(_outputHelper);
+
+            Task ruleCompletedTask = null;
+
+            await ScenarioRunner.SingleTarget(_outputHelper,
+                _httpClientFactory,
+                DiagnosticPortConnectionMode.Listen,
+                TestAppScenarios.TraceEvents.Name,
+                appValidate: async (appRunner, apiClient) =>
+                {
+                    await appRunner.SendCommandAsync(TestAppScenarios.TraceEvents.Commands.EmitUniqueEvent);
+                    await ruleCompletedTask;
+                    await appRunner.SendCommandAsync(TestAppScenarios.TraceEvents.Commands.ShutdownScenario);
+                },
+                configureTool: runner =>
+                {
+                    runner.ConfigurationFromEnvironment.AddFileSystemEgress(EgressProvider, tempDirectory.FullName);
+                    runner.ConfigurationFromEnvironment.CreateCollectionRule(DefaultRuleName)
+                        .SetStartupTrigger()
+                        .AddCollectTraceAction(
+                            new EventPipeProvider[] {
+                                new EventPipeProvider()
+                                {
+                                    Name = EventProviderName,
+                                    Keywords = "-1"
+                                }
+                            },
+                            EgressProvider, options =>
+                            {
+                                options.Duration = duration ?? TimeSpan.Parse(ActionOptionsConstants.Duration_MaxValue);
```

Comment:
Let's use a duration that is shorter than collection rule completion timeout but long enough where the positive validation tests can detect the stopping event.